### PR TITLE
Deprecate `adjust_infection_to_report()`, `report_cases()`, and `sample_approx_dist()`

### LINKED ISSUE #520: Deprecate `report_cases`, `sample_approx_dist` and `adjust_infection_report`
As part of this we could probably deprecate `sample_approx_dist` and `adjust_infection_report`, and replace `dist_skel` with a simpler version that just calculates the pmf in line with #411

_Originally posted by @sbfnk in https://github.com/epiforecasts/EpiNow2/issues/504#issuecomment-1810697905_

### PR DESCRIPTION
**Pull request by jamesmbaazam**
## Description

This PR closes #520.

The functions `sample_approx_dist()`, `report_cases()`, and `adjust_infection_reports()` have been deprecated as the functionality they provide can now be achieved with `simulate_secondary()`.

## Initial submission checklist

- [x] My PR is based on a package issue and I have explicitly linked it.
- [x] I have tested my changes locally (using `devtools::test()` and `devtools::check()`).
- [x] I have added or updated unit tests where necessary.
- [x] I have updated the documentation if required and rebuilt docs if yes (using `devtools::document()`).
- [x] I have followed the established coding standards (and checked using `lintr::lint_package()`).
- [x] I have added a news item linked to this PR.

## After the initial Pull Request 

- [ ] I have reviewed Checks for this PR and addressed any issues as far as I am able.

### CODE REVIEW

File: R/adjust.R
Comment (sbfnk) [inline R/adjust.R:86]: why nolint here? Don't see the need to have a missing argument.

File: R/adjust.R
Comment (sbfnk) [inline R/adjust.R:89]: can we add a detail that this will be completely removed in 2.0.0 as for the other deprecations (see examples in https://github.com/epiforecasts/EpiNow2/blob/dist-interface/R/deprecated.R).

File: R/dist.R
Comment (sbfnk) [inline R/dist.R:710]: same comments as above

File: R/report.R
Comment (sbfnk) [inline R/report.R:55]: same comments as above

File: tests/testthat/test-adjust_infection_to_report.R
Comment (sbfnk) [inline tests/testthat/test-adjust_infection_to_report.R:42]:
```suggestion
test_that("deprecated functions are deprecated", {
```

File: tests/testthat/test-dist.R
Comment (sbfnk) [inline tests/testthat/test-dist.R:30]:
```suggestion
test_that("deprecated functions are deprecated", {
```

File: tests/testthat/test-report_cases.R
Comment (sbfnk) [inline tests/testthat/test-report_cases.R:23]:
```suggestion
test_that("deprecated functions are deprecated", {
```

**Review by sbfnk — CHANGES_REQUESTED** [review]
Good stuff.

I think it would be good to point the user to how they would `simulate_secondary` to reproduce the functionality of the deprecated functions - but that should wait until #504 is merged as otherwise the syntax will have to be updated again.

I'd suggest to address the comments/issues and then convert to a draft PR to revisit after #504 is merged.

File: NEWS.md
Comment (sbfnk) [inline NEWS.md:12]:
```suggestion
* The functions `sample_approx_dist()`, `report_cases()`, and `adjust_infection_reports()` have been deprecated as the functionality they provide can now be achieved with `simulate_secondary()`. See #597 by @jamesmbaazam and reviewed by @sbfnk.
```

File: R/adjust.R
Comment (jamesmbaazam) [reply R/adjust.R]: Thanks. I've done that here bad9ea4aa848670a479abc7733dbe823362892ea.

File: R/adjust.R
Comment (jamesmbaazam) [reply R/adjust.R]: That's a good point. I initially thought it was because `lintr` thought I was calling a function without arguments but turns out it was the trailing comma. I've removed the trailing comma and removed the tags in 42c9f84df7d08c74354f8781bfd7e909d55e57ab and 5dfe07d065cbd3e8de0d9fe3a5194847a1261609.

File: R/report.R
Comment (jamesmbaazam) [reply R/report.R]: This has been removed.

File: R/dist.R
Comment (jamesmbaazam) [reply R/dist.R]: This has been fixed.

**Comment by jamesmbaazam** [comment]
> Good stuff.
> 
> I think it would be good to point the user to how they would `simulate_secondary` to reproduce the functionality of the deprecated functions - but that should wait until #504 is merged as otherwise the syntax will have to be updated again.
> 
> I'd suggest to address the comments/issues and then convert to a draft PR to revisit after #504 is merged.

Thanks for the review. I'll convert the PR to a draft now until #504 is merged.

File: R/adjust.R
Comment (sbfnk) [inline R/adjust.R:89]: this should be in a named argument `details` (see failing tests).

Comment (jamesmbaazam) [reply R/adjust.R]: My bad. I've fixed it now.

**Comment by sbfnk** [comment]
@jamesmbaazam I've now merged main which seems to pass tests. I think once we add some guidance to the user this is good to go.

Not quite sure where this best lives but here is some code for `adjust_infection_to_report`:
```r
cases <- data.table::copy(example_confirmed)[, cases := as.integer(confirm)]
delay_def <- lognorm_dist_def(
  mean = 5, mean_sd = 1, sd = 3, sd_sd = 1,
  max_value = 30, samples = 1, to_log = TRUE
)
report <- adjust_infection_to_report(
  cases, delay_defs = delay, reporting_model = function(n) rpois(length(n), n)
)
print(report)


cases <- data.table::copy(example_confirmed)[, primary := as.integer(confirm)]
uncertain_delay <- LogNormal(mean = Normal(5, 1), sd = Normal(3, 1), max = 30)
delay <- fix_dist(uncertain_delay, strategy = "sample")
report <- simulate_secondary(
  cases, delays = delay_opts(delay), obs = obs_opts(family = "poisson")
)
print(report)
```

and `report_cases`:
```r
cases <- example_confirmed[1:40]
cases <- cases[, cases := as.integer(confirm)]
cases <- cases[, confirm := NULL][, sample := 1]
reported_cases <- report_cases(
  case_estimates = cases,
  delays = delay_opts(example_incubation_period + example_reporting_delay),
  type = "sample"
)
print(reported_cases$samples)

cases <- example_confirmed[1:40]
cases <- cases[, primary := as.integer(confirm)]
report <- simulate_secondary(
  cases,
  delays = delay_opts(
    fix_dist(example_incubation_period + example_reporting_delay)
  ),
  obs = obs_opts(family = "poisson")
)
print(report)
```

**Comment by jamesmbaazam** [comment]
> @jamesmbaazam I've now merged main which seems to pass tests. I think once we add some guidance to the user this is good to go.
> 
> Not quite sure where this best lives but here is some code for `adjust_infection_to_report`:
> 
> ```r
> cases <- data.table::copy(example_confirmed)[, cases := as.integer(confirm)]
> delay_def <- lognorm_dist_def(
>   mean = 5, mean_sd = 1, sd = 3, sd_sd = 1,
>   max_value = 30, samples = 1, to_log = TRUE
> )
> report <- adjust_infection_to_report(
>   cases, delay_defs = delay, reporting_model = function(n) rpois(length(n), n)
> )
> print(report)
> 
> 
> cases <- data.table::copy(example_confirmed)[, primary := as.integer(confirm)]
> uncertain_delay <- LogNormal(mean = Normal(5, 1), sd = Normal(3, 1), max = 30)
> delay <- fix_dist(uncertain_delay, strategy = "sample")
> report <- simulate_secondary(
>   cases, delays = delay_opts(delay), obs = obs_opts(family = "poisson")
> )
> print(report)
> ```
> 
> and `report_cases`:
> 
> ```r
> cases <- example_confirmed[1:40]
> cases <- cases[, cases := as.integer(confirm)]
> cases <- cases[, confirm := NULL][, sample := 1]
> reported_cases <- report_cases(
>   case_estimates = cases,
>   delays = delay_opts(example_incubation_period + example_reporting_delay),
>   type = "sample"
> )
> print(reported_cases$samples)
> 
> cases <- example_confirmed[1:40]
> cases <- cases[, primary := as.integer(confirm)]
> report <- simulate_secondary(
>   cases,
>   delays = delay_opts(
>     fix_dist(example_incubation_period + example_reporting_delay)
>   ),
>   obs = obs_opts(family = "poisson")
> )
> print(report)
> ```

We could add these to the example files of the deprecated functions and add a details argument to `deprecate_warn()` to point users to it., i.e., "Use `simulate_secondary()` instead. See ?report_cases for equivalent usage. Note that this <deprecated function> will be removed in v2.0.0."

**Review by sbfnk — APPROVED** [review]

**Comment by sbfnk** [comment]
Great work!

**Pull request merged**
